feat: argument passing 구현

hdh853 opened the pull request:
## 🔗 관련 이슈
<!-- 이슈 번호를 연결해주세요 -->
closes #20 

## 📌 작업 내용
### 구현 사항
<!-- 구현내용에 대해서 설명해주세요. -->
<!-- 예시 :
- 이슈 템플릿, PR 템플릿 작성
- READEME 파일 최초 작성
 --> 
- load함수 내부에서 인자전달 구현

### 변경된 파일
<!-- 구현 내용으로 인해 변경된 파일에 대해 설명해주세요 -->
<!-- 예시:
- README.md
  - readme 파일 최초 생성
- issue_template.md
  - 이슈 템플릿 최초 생성
-->
- process.c
  - load에 인자로 넘어온 file_name을 malloc으로 할당한 arg_v에 쪼개어 저장
  - 저장한 인자들을 argvP에 저장하며 유저스택에 쌓고 8바이트 패딩 맞춤
  - argvP의 역순으로 유저스택에 저장
  - fake return address 추가
  - rdi, rsi에 arg_c 값과 argv의 엔트리 포인트 저장 


### 추가 사항
<!-- 트러블 슈팅 등 추가로 설명할 내용이 있으시다면 아래에 작성해주세요. -->

## ✅ 셀프 체크리스트
<!-- 체크리스트를 모두 완료한 후 제출해주세요 -->
<!-- 완료 처리를 표시하려면 [x], 완료되지 않았다면 [ ] -->
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 주석이나 디버깅 코드를 제거했나요?
- [ ] 관련 테스트가 모두 통과하나요?
- [x] 커밋 메시지가 컨벤션에 맞나요?

hdh853 committed:
style: init.c, process.c 포멧팅 적용

hdh853 committed:
feat: argument passing 구현

hyunmuam commented on userprog/process.c:
```diff
@@ -404,13 +428,49 @@ static bool load(const char* file_name, struct intr_frame* if_)
+    while (((uint8_t*)USER_STACK - sp) % 8 != 0) {
+        sp--;
+        *sp = 0;
+    }
```
이 부분은 포인터 연산으로 개선할 수 있을 것 같습니다.
현재는 최대 7번 루프를 도는데, 비트 연산으로 한 번에 계산하면 더 간결하고 빠를 거에요.

hdh853 replied:
그러네요. '매 포인터를 찍어 거기를 0값을 넣어야 한다' 해서 일일이 반복문 돌아야한다 생각했는데, memcpy를 사용한다면 반복문을 돌지 않고도 해결할 수 있겠네요.

hyunmuam replied:
@hdh853 
넵 조금 더 자세히 말씀드리자면
`palloc_get_page(PAL_ZERO)`로 할당받으면 이미 페이지가 0으로 초기화되어 있으므로, `memcpy`로 문자열을 복사한 후 남은 공간은 이미 0입니다. 따라서 while 루프로 바이트 단위로 0을 쓸 필요 없이, 포인터 연산으로 정렬된 주소만 계산하면 될 것 같습니다.

hyunmuam commented on userprog/process.c:
```diff
@@ -332,6 +336,26 @@ static bool load(const char* file_name, struct intr_frame* if_)
+    while (arg_v[arg_c] != NULL && arg_c < LOADER_ARGS_LEN) {
+        arg_c++;
+        arg_v[arg_c] = strtok_r(NULL, " ", &save_ptr);
+    }
```
`strtok_r`은 원본 문자열을 수정하는 함수입니다. 
현재 구현에서 `file_name`을 직접 파싱하면, `load()` 종료 후 
`process_exec()`의 `palloc_free_page(file_name)`에서 
이미 훼손된 문자열을 참조하게 될텐데 문제가 없을까요?

hdh853 replied:
현재 코드상에선 아래에 file_name을 갱신해주는 로직이 있어서 동작에는 문제가 없을것 같습니다. 다만 strtok_r을 사용해 원본을 훼손하면서 작업하는게 맞는가? 는 고민을 해보겠습니다.

hyunmuam replied:
@hdh853 
> 현재 코드상에선 아래에 file_name을 갱신해주는 로직이 있어서 동작에는 문제가 없을것 같습니다.

어디에 있는 로직인지 조금 더 자세히 말씀해주실 수 있나요?

hyunmuam commented on userprog/process.c:
```diff
@@ -332,6 +336,26 @@ static bool load(const char* file_name, struct intr_frame* if_)
+    }
+
+    // 파일명 설정
+    file_name = arg_v[0];
+    if (file_name == NULL) {
+        printf("load: no file name\n");
+        goto done;
+    }
```
이 출력문(`load: no file name`)이 스펙에 명시된 내용인지 확인 부탁드립니다.
만약 디버깅용으로 넣으신 거라면, 자동 채점 시스템에서 오류를 일으킬 수 있으니 제거하는 것이 안전할 것 같습니다.

hdh853 replied:
그러네요 스펙에 명시된게 아니라 그냥 생각없이 넣은 printf 문인데, goto done; 만 해주는게 맞을것 같네요

hyunmuam commented on userprog/process.c:
```diff
@@ -404,13 +428,49 @@ static bool load(const char* file_name, struct intr_frame* if_)
+        memcpy(sp, arg_v[i], len); // sp에 arg_v[i]를 len만큼 덮어씀
+
+        argvP[i] = (char*)sp;
+    }
```
`argvP` 배열의 역할이 무엇인가요?

만약 현재 구현에서는 스택에 복사된 문자열 주소를 저장하기 위해 
별도 배열을 사용하고 있다면, 이는 커널 스택(현재 `userprog`기준 4KB)에서 오버플로우 위험을 높일 수 있다고 생각합니다.

원본 `arg_v` 배열을 직접 갱신하면 
추가 메모리 할당 없이 동일한 결과를 얻을 것 같습니다.

hdh853 replied:
argvP 배열의 역할은 정확히 생각하고 계신대로 쓰이는게 맞습니다. 
말씀하신 방식이 argv에서 원소 값을 넣을때 마다 그당시의 sp 주소를 갱신하고, 그후 포인터를 넣을 때 갱신된 원소값의 sp 주소를 넣는 식인거죠? 그때는 생각못했는데, 이렇게 하면 argP 없이도 가능할 것 같네요. 감사합니다!

hdh853 committed:
refactor: 피드백 반영
- 인자로 받은 file_name 복사해서 사용
- argvP 대신 arg_v 갱신해가며 사용
- 패딩 채우는 while문 사용x. memset 사용

hyunmuam commented on userprog/process.c:
```diff
@@ -440,19 +443,20 @@ static bool load(const char* file_name, struct intr_frame* if_)
+    size_t pad = (8 - (used % 8)) % 8;
+
+    sp -= pad;
+    memset(sp, 0, pad);
```
현재 구현도 동작하지만, 비트 연산을 사용하면 더 효율적으로 처리할 수 있습니다.
**현재 방식 (산술 연산)**
```
size_t used = (uint8_t*)USER_STACK - sp; // 뺄셈 1회
size_t pad = (8 - (used % 8)) % 8; // 나눗셈(mod) 2회, 뺄셈 1회
sp -= pad; // 뺄셈 1회
memset(sp, 0, pad); // 함수 호출 + 루프
```
→ 총 5회 이상의 연산 + 함수 호출

**개선 방식 (비트 연산)**
```
sp = (uint8_t*)((uintptr_t)sp & ~0x7UL); // AND 연산 1회
```
→ 단 1회의 연산

### 원리 설명
8바이트 정렬이란 "주소의 하위 3비트가 000"인 상태입니다.
| 주소 예시 | 이진수 | 8의 배수? |
|-----------|--------|-----------|
| 0x4747fff5 | ...1111 0**101** | ❌ |
| 0x4747fff0 | ...1111 0**000** | ✅ |

`~0x7`은 `...11111000` (하위 3비트만 0)이므로,
`sp & ~0x7`을 하면 하위 3비트가 강제로 0이 됩니다:
```
sp = 0x4747fff5 = ...1111 0101
& ~0x7 = ...1111 1000
─────────────────────────────────────
결과 = 0x4747fff0 = ...1111 0000 ← 8의 배수!
```
### memset이 필요 없는 이유
`palloc_get_page(PAL_ZERO)`로 할당받은 페이지는 이미 전체가 0으로 초기화되어 있으므로,
패딩 영역에 다시 0을 쓸 필요가 없습니다.

### 결론
```
// Before
size_t used = (uint8_t*)USER_STACK - sp;
size_t pad = (8 - (used % 8)) % 8;
sp -= pad;
memset(sp, 0, pad);

// After (권장)
sp = (uint8_t*)((uintptr_t)sp & ~0x7UL);
```
비트 연산은 CPU에서 단일 사이클로 처리되고, 나눗셈(mod)은 수십 사이클이 소요됩니다.
